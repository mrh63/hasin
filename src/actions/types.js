export const actionTypes = {
    //PRODUCTS
    ADD_PRODUCT: 'ADD_PRODUCT',
    REMOVE_PRODUCT: 'REMOVE_PRODUCT',
    SET_PRODUCTS: 'SET_PRODUCTS',
};