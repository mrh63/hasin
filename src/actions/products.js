import { actionTypes } from './types';
import axios from 'axios';


//REMOVE_PRODUCT
export const removeProduct = ({id} = {}) => ({
    type: actionTypes.REMOVE_PRODUCT,
    id
});

//SET_PRODUCTS
export const setProducts = (products) => ({
    type: actionTypes.SET_PRODUCTS,
    products
});

export const startSetProducts = (removeItem = []) => async dispatch => {

    const products = [];

    const res = await axios.get(['http://127.0.0.1:3000', 'data'].join('/'))
    .catch(function(thrown) {
        console.log('server not available');
        return dispatch(setProducts(products));
    });

    for (let key in res.data) {
        if (removeItem.indexOf(key) === -1)
            products.push({productKey: key, ...res.data[key]});
    }
    dispatch(setProducts(products));
}