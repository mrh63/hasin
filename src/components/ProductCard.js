import React from 'React';
import { useDispatch } from 'react-redux';
import { Card, CardTitle, CardBody, CardText, CardLink } from 'reactstrap';
import { NavLink as RRNavLink } from 'react-router-dom';
import { removeProduct } from '../actions/products';

const ProductCard = ({productKey, id, title, subtitle, image}) => {

    const dispatch = useDispatch();

    const onRemove = () => {
        const removedProducts = localStorage.removedProducts ? JSON.parse(localStorage.removedProducts) : [];
        removedProducts.push(productKey);
        const json = JSON.stringify(removedProducts)
        localStorage.setItem('removedProducts', json);
        dispatch(removeProduct({id}));
    }

    return (
        <div className="product-card">
            <Card className="h-100">
                <CardBody>
                    <CardTitle>{title}</CardTitle>
                    </CardBody>
                        <img width="100%" src={image} alt="Card cap" />
                    <CardBody>
                    <CardText className="text-justify">{subtitle}</CardText>
                    <CardLink href="#" onClick={onRemove}> حذف </CardLink>
                    <CardLink to={`/${title}`} tag={RRNavLink} className="mr-5"> جزییات بیشتر </CardLink>
                </CardBody>
            </Card>
        </div>
    )
}

export default ProductCard;