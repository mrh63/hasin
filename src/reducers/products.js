import { actionTypes } from '../actions/types';

//Products Reducer
const productsReducerDefaultState = [];

export default (state = productsReducerDefaultState, action) => {
    switch (action.type) {
        case actionTypes.REMOVE_PRODUCT:
            return state.filter(({id}) => id !== action.id);
        case actionTypes.SET_PRODUCTS:
            return action.products;
        default:
            return state;
    }
};