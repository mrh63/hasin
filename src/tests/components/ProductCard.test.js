import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from 'react-redux'
import { mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import ProductCard from '../../components/ProductCard';
import products from '../fixtures/products';
import { removeProduct } from '../../actions/products';

const createMockStore = configureMockStore([thunk]);

let store, wrapper;
beforeEach(() => {
    store = createMockStore({
        products
    });
    store.dispatch = jest.fn();
    wrapper = mount(
        <Provider store={store}>
            <Router>
                <ProductCard {...products[2]} />
            </Router>
        </Provider>
    )
});

test('should render ProductCard correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should dispatch the correct action on remove button click', () => {
    wrapper.find('CardLink[href="#"]').simulate('click');
    expect(store.dispatch).toHaveBeenLastCalledWith(removeProduct({id:products[2].id}));
});
