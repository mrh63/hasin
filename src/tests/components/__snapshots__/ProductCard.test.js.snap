// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProductCard correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <ProductCard
        description="فروشگاه اینترنتی گوشی‌شاپ در سال ۱۳۸۴ به عنوان اولین وب‌سایت اطلاع‌رسانی و فروشگاه تخصصی گوشی موبایل آغاز به‌کار کرد و افتخار دارد که با بهره‌مندی از تیم متخصص و متبحر نرم‌افزاری، اولین پلت‌فرم فروشگاهی هوشمند کاملا ایرانی را بنیان نهاده است.         کسب نماد اعتماد دو ستاره از مرکز توسعه تجارت الکترونیک، حضور در هیات موسس و عضویت در هیات مدیره انجمن صنفی کسب‌وکار اینترنتی و موفقیت در استمرار حضور ده‌ساله، مرهون اعتماد کاربران محترم است و گوشی‌شاپ قدردان این اعتماد بوده و به‌عنوان یک فروشگاه اینترنتی رعایت جانب انصاف و صداقت به عنوان اصلی‌ترین پایه این اعتماد را سرلوحه کار خود می‌داند و همواره سعی در گسترش خدمات خود خواهد داشت."
        id={3}
        image="/images/gooshishop.jpg"
        productKey="gooshishop"
        subtitle="فروشگاه اینترنتی گوشی‌شاپ در سال ۱۳۸۴ به عنوان اولین وب‌سایت اطلاع‌رسانی و فروشگاه تخصصی گوشی موبایل آغاز به‌کار کرد."
        title="گوشی شاپ"
      >
        <div
          className="product-card"
        >
          <Card
            className="h-100"
            tag="div"
          >
            <div
              className="h-100 card"
            >
              <CardBody
                tag="div"
              >
                <div
                  className="card-body"
                >
                  <CardTitle
                    tag="div"
                  >
                    <div
                      className="card-title"
                    >
                      گوشی شاپ
                    </div>
                  </CardTitle>
                </div>
              </CardBody>
              <img
                alt="Card cap"
                src="/images/gooshishop.jpg"
                width="100%"
              />
              <CardBody
                tag="div"
              >
                <div
                  className="card-body"
                >
                  <CardText
                    className="text-justify"
                    tag="p"
                  >
                    <p
                      className="text-justify card-text"
                    >
                      فروشگاه اینترنتی گوشی‌شاپ در سال ۱۳۸۴ به عنوان اولین وب‌سایت اطلاع‌رسانی و فروشگاه تخصصی گوشی موبایل آغاز به‌کار کرد.
                    </p>
                  </CardText>
                  <CardLink
                    href="#"
                    onClick={[Function]}
                    tag="a"
                  >
                    <a
                      className="card-link"
                      href="#"
                      onClick={[Function]}
                    >
                       حذف 
                    </a>
                  </CardLink>
                  <CardLink
                    className="mr-5"
                    tag={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "NavLink",
                        "propTypes": Object {
                          "activeClassName": [Function],
                          "activeStyle": [Function],
                          "aria-current": [Function],
                          "className": [Function],
                          "exact": [Function],
                          "innerRef": [Function],
                          "isActive": [Function],
                          "location": [Function],
                          "onClick": [Function],
                          "replace": [Function],
                          "strict": [Function],
                          "style": [Function],
                          "target": [Function],
                          "to": [Function],
                        },
                        "render": [Function],
                      }
                    }
                    to="/گوشی شاپ"
                  >
                    <NavLink
                      className="mr-5 card-link"
                      to="/گوشی شاپ"
                    >
                      <Link
                        aria-current={null}
                        className="mr-5 card-link"
                        to={
                          Object {
                            "hash": "",
                            "pathname": "/گوشی شاپ",
                            "search": "",
                            "state": null,
                          }
                        }
                      >
                        <LinkAnchor
                          aria-current={null}
                          className="mr-5 card-link"
                          href="/گوشی شاپ"
                          navigate={[Function]}
                        >
                          <a
                            aria-current={null}
                            className="mr-5 card-link"
                            href="/گوشی شاپ"
                            onClick={[Function]}
                          >
                             جزییات بیشتر 
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </CardLink>
                </div>
              </CardBody>
            </div>
          </Card>
        </div>
      </ProductCard>
    </Router>
  </BrowserRouter>
</Provider>
`;
